avocado/core/jobdata.py

```diff
@@ -0,0 +1,217 @@
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+# Copyright: Red Hat Inc. 2016
+# Author: Amador Pahim <apahim@redhat.com>
+
+"""
+Record/retrieve job information
+"""
+
+import ast
+import glob
+import os
+import pickle
+import sys
+
+
+from . import exit_codes
+from .settings import settings
+from ..utils.path import init_dir
+
+
+JOB_DATA_DIR = 'jobdata'
+JOB_DATA_FALLBACK_DIR = 'replay'
+CONFIG_FILENAME = 'config'
+URLS_FILENAME = 'urls'
+MUX_FILENAME = 'multiplex'
+PWD_FILENAME = 'pwd'
+ARGS_FILENAME = 'args'
+CMDLINE_FILENAME = 'cmdline'
+
+
+def record(args, logdir, mux, urls=None, cmdline=None):
+    """
+    Records all required job information.
+    """
+    base_dir = init_dir(logdir, JOB_DATA_DIR)
+    path_cfg = os.path.join(base_dir, CONFIG_FILENAME)
+    path_urls = os.path.join(base_dir, URLS_FILENAME)
+    path_mux = os.path.join(base_dir, MUX_FILENAME)
+    path_pwd = os.path.join(base_dir, PWD_FILENAME)
+    path_args = os.path.join(base_dir, ARGS_FILENAME)
+    path_cmdline = os.path.join(base_dir, CMDLINE_FILENAME)
+
+    if urls:
+        with open(path_urls, 'w') as urls_file:
+            urls_file.write('%s' % urls)
+
+    with open(path_cfg, 'w') as config_file:
+        settings.config.write(config_file)
+
+    with open(path_mux, 'w') as mux_file:
+        pickle.dump(mux, mux_file, pickle.HIGHEST_PROTOCOL)
```

Comment:
There are 3 protocols currently. If you say nothing, it will use the `v0`, which is the original ASCII protocol. The `HIGHEST_PROTOCOL` is the `v2`, the same since python 2.3. As far as you use at least python 2.3 in all machines, this is not expected to have compatibility issues. New versions of pickle are not expected to show up in the python versions we support currently.